{% extends "base_generic.html" %}
{% load static %}
{% block page-title %}<title>About this site</title>{% endblock %}
{% block title %}<div class="logo"><h2>About</h2></div>{% endblock %}

{% block content %}
        <section id="one" class="wrapper style3">
          <div class="inner">
            <div class="row 200%">
              <div class="6u 12u$(medium)">
                <h3>About this site{%if site_title %}: {{ site_title }}{% else %}{% endif %}</h3>
                <h5>This is a microbial genome database based on the <a href="https://github.com/aekazakov/genome-depot">GenomeDepot platform</a>. GenomeDepot is a tool for annotation, management and comparative analysis of prokaryotic genomes.</h5>
                <h5>The database aggregates gene function annotations imported from GenBank-formatted genome files and generated by the GenomeDepot annotation pipeline. The pipeline supports various computational tools for function annotation, orthology assignment and protein domain mapping (see the full list of tools in the <a href="https://aekazakov.github.io/genome-depot/">GenomeDepot documentation website</a>).</h5>
                <h3>Image credits</h3>
                <h5>Background image: <i>B. burgdorferi</i> bacteria. Photo by <a href="https://pixnio.com/science/microscopy-images/borrelia-burgdorferi/b-burgdorferi-bacteria">Jamice Haney Carr, Claudia Molins, USCDCP</a> on <a href="https://pixnio.com/">Pixnio</a></h5>
                <h5>Start page image: <i>Streptococcus pneumoniae</i> bacterial colonies that were grown on primary isolation medium. Photo by <a href="https://pixnio.com/science/microscopy-images/streptococcus-pneumoniae/streptococcus-pneumoniae-bacterial-colonies-that-were-grown-on-primary-isolation-medium">Dr. Richard Facklam, USCDCP</a> on <a href="https://pixnio.com/">Pixnio</a></h5>
                <h5>Earth spinning rotating animation. <a href="https://commons.wikimedia.org/wiki/File:Earth-spinning-rotating-animation-40.gif">Amirabbaszakavi</a>, <a href="https://creativecommons.org/licenses/by-sa/4.0">CC BY-SA 4.0</a>, via Wikimedia Commons</h5>
                <h3>Acknowledgement</h3>
                <h5>This material by ENIGMA- Ecosystems and Networks Integrated with Genes and Molecular Assemblies (<a href="http://enigma.lbl.gov">http://enigma.lbl.gov</a>), a Science Focus Area Program at Lawrence Berkeley National Laboratory is based upon work supported by the U.S. Department of Energy, Office of Science, Office of Biological & Environmental Research under contract number DE-AC02-05CH11231</h5>
              </div>
              <div class="6u$ 12u$(medium)">
                <h3>Database statistics</h3>
                <div class="align-center" id="waitbox">Retrieving statistics...<br><img src="{% static 'images/spinner.gif' %}"></div>
                <div id="stats">
                </div>
                <div id ="error-box" class="error box"></div>
              </div>
            </div>
            <hr/>
            <div class="grid-style">
              <div class="content">
                <h3>Quick help</h3>
                <h5>This section contains few key concepts and short description of GenomeDepot capabilities. For detailed help, consult <a href = "https://aekazakov.github.io/genome-depot/user">User guide page</a> of the GenomeDepot documentation website at <a href="https://aekazakov.github.io/genome-depot">https://aekazakov.github.io/genome-depot/user</a>.</h5>
                <h3>Genomes</h3>
                <h5>A genome is a complete or partial assembly of genomic sequences of a particular organism. Genome can include one or several contiguous nucleotide sequences, or contigs.</h5><h5><a href="{% url 'genome_list' %}">List of all genomes</a> is available from the start page or from the site menu. Click on the name of a genome in the list to open genome page. A genome page has links to the list of genes, interactive genome browser and a number of fields for search of genes by functional annotations.</h5>
                <h3>Strains</h3>
                <h5>Genomes of organisms cultivated under controlled laboratory conditions are linked to strain entries. <a href="{% url 'strain_list' %}">Full list of strains</a> is available from the start page or from the site menu. Click on the name of a strain in the list to open strain details page. The strain details page contains links to genome assemblies for the strain and strain metadata.</h5>
                <h3>Samples</h3>
                <h5>Genomes of uncultured organisms assembled from metagenomic sequences are linked to sample entries. Typically, several metagenome-assembled genomes (MAGs) are reconstructed from one sample. <a href="{% url 'sample_list' %}">Full list of samples</a> is available from the start page or from the site menu. Click on the name of a sample in the list to open the sample page. A sample details page has links to genome assemblies linked to the sample and may contain some strain metadata.</h5>
                <h3>Genes</h3>
                <h5>GenomeDepot database is gene-centric, thus gene pages is your primary destination point. At the top of a gene page, there are genome name, gene coordinates and gene function that came with the genome file. Next to them, for coding genes you can find links to EggNOG family pages and genome neighborhood comparative viewer.</h5>
                <h5>Under the gene information block, you will see interactive genome browser showing that gene (highlighted) and surrounding areas. Click on a gene in the interactive browser opens a pop-up with a link to the gene's page, nucleotide sequence of the gene and some details about the gene.</h5>
                <h5>The next block shows functional annotations for coding genes. The data in the left column were generated by EggNOG-mapper (like KEGG mappings, EC number, COG class, GO terms etc.). Annotations in the right column were generated by other annotation tools.</h5>
                <h5>At the bottom of the page you can find links to external tools for sequence analysis and amino acid sequence (only for coding genes).</h5>
                <h5>There are several ways to find a gene of interest in GenomeDepot. You can search genes by name or function from search form at the top of that page. Or, you can open a list of genes for a genome of interest from the genome page. Finally, you can search by functional annotation or by nucleotide or amino acid sequence.</h5>
              </div>
             <div class="content">
                <h3>Comparative gene neighborhood viewer</h3>
                <h5>Click on one of "Show genome neighborhood conservation" links on a gene page to open the comparative viewer. It shows gene neighborhood around members of the selected orthologous group. The starting gene is always in the top row, other rows show most similar members of the orthologous group.</h5>
                <h5>Click on a gene name opens a page for the gene. Genes on the plot are colored according to the orthology assignment by EggNOG-mapper (usually, by COGs). If a gene was not assigned to any OGs, it is colored grey.</h5>
                <h5>A combined functional profile of all genes in the neighborhood is shown under the genomes plot.</h5>
                <h3>KEGG map viewer</h3>
                <h5>If the Functional annotations section of the gene page contains links to KEGG pathways, these links open a page with the list of all genes in the current genome mapped to that pathway. If there is a link named "View KEGG map for these functions" at the top of the page, a click on the link opens a page with a list of genes for that map. On the right side, a KEGG map is loaded from the KEGG website. Genes in the list and on the map are colored by KEGG Orthology assignment. Please note that this functionality depends on external web resource.</h5>
                <h3>Search page</h3>
                <h5>Open <a href="{% url 'textsearch' %}">text search page</a> to search for genes by name, by annotated function or by one of functional categories (like KEGG pathways or EC numbers). You can enter an identifier or keywords in the search form to filter list of categories, or click a link next to the search form to browse a list of entries in that category. Then click on the name of a category to open list of all genes assigned to that category</h5>
                <h3>Search by protein or nucleotide sequence</h3>
                <h5>The search application uses BLASTP and and megablast for sequence similarity search. Open <a href="{% url 'proteinsearch' %}">protein sequence search</a> or <a href="{% url 'nucleotidesearch' %}">nucleotide sequence search</a> page from the start page or from the site menu. Enter your sequence in the FASTA format in the search form and click "Start search". The Query must contain one sequence only. The query sequence can have FASTA header.</h5>
                <h5>Sequence search can take some time depending on the availability of computational resources, so please do not reload the page.</h5>
                <h5>On the protein search results page, a table of hits shows target genes (with link to the gene page), amino acid identity %, length of alignment, E-value and bit-score.</h5>
                <h5>On the nucleotide search results page, a table of hits shows target genomes (with link to the genome page), hit identity %, length of alignment, E-value and bit-score. Click on a genome name opens the genome page with the highlighted hit area in the interactive browser.</h5>
              </div>
            </div>
          </div>
        </section>
<script src="//code.jquery.com/jquery-1.11.3.min.js"></script>
<script>
    $(document).ready(function(){
        $.ajax({
            url: "{% url 'getstats' %}",
            method: 'GET',
            datatype: 'json',
            data: {},
            xhrFields: {
            responseType: 'text'
            },
            beforeSend: function() {
            $("#waitbox").show();
            $("#export-profile").hide();
            },
            success: function (data) {
            $("#waitbox").hide();
            $("#stats").append(data["stats"])
            },
            error: function(XMLHttpRequest, textStatus, errorThrown){
            $("#waitbox").hide();
            $("#error-box").append(textStatus + ": " + errorThrown);
            }
        });
    });
</script>
        
{% endblock %}
